### PR TITLE
Navigation: PoC for changing nav menu position

### LINKED ISSUE #13573: Sidebar -> Top nav bar
Side menu cons
* Taking up valuable horizontal space. 
* It's creating a sort of half framing (together with dashboard nav) where we have side & top bar that does not look great
* Auto hide removes sidebar which changes the width of the dashboard which requires all panels to update & re-render
* No nice place to put home icon & search input

### Proposed new top nav:
image

The above is just a mockup, not a final design. Need to add an explore icon, and a create button as well (see trello for reference). And change order of help icon and profile avatar image.

### Slide-out menu
I image the menu button would open up a slide in (over page) menu. Look of side menu needs a design mockup. 

### Hiding top menu
To save precious vertical space when in panel edit mode I think we need to hide top menu when entering panel edit/fullscreen mode. 

Of course it will hide like the sidemenu does in inactive & tv & kiosk modes.

### PR DESCRIPTION
**Pull request by ashharrison90**
**What this PR does / why we need it**:

- adds a new `Navigation position` option to the Shared preferences page behind the `newNavigation` feature toggle:
image
  - default is `left`: can add support to change this later in the configuration
  - this works exactly the same as the theme. it can be set at an org and user level. user level takes precedence if both are set.
- adds support in the api calls for the `navPosition` property. this currently isn't feature toggled, but probably should be. it's only because i don't really know what i'm doing with the go backend yet 😂 
- some screenshots:
image
image
image
image

**Which issue(s) this PR fixes**:

For https://github.com/grafana/grafana/issues/37961
Fixes https://github.com/grafana/grafana/issues/13573

**Special notes for your reviewer**:
- to test:
  - set the feature toggle in `custom.ini`:
  ```
  [feature_toggles]
  enable = newNavigation
  ```
  - **take a copy of your grafana.db** (in `data/grafana.db`). once you've enabled the feature toggle and ran `make run` on this branch, it will migrate the db with a new value (`nav_position`) and you won't be able to go back. so take a copy, then when you're done testing you can go turn the toggle off, replace the db with your backup and return to your old branch safely.
- my go is bad and i find the grafana backend a little confusing 😅  so there's probably loads of improvements to be made there

### CODE REVIEW

File: pkg/services/sqlstore/migrations/preferences_mig.go [hunk 1]
```diff
@@ -15,6 +15,7 @@ func addPreferencesMigrations(mg *Migrator) {
 			{Name: "home_dashboard_id", Type: DB_BigInt, Nullable: false},
 			{Name: "timezone", Type: DB_NVarchar, Length: 50, Nullable: false},
 			{Name: "theme", Type: DB_NVarchar, Length: 20, Nullable: false},
+			{Name: "nav_position", Type: DB_NVarchar, Length: 20, Nullable: false},
```
Comment (hugohaggmark) [inline pkg/services/sqlstore/migrations/preferences_mig.go:18]: Aha, I see what you did here, the migrations are unique so you can't add the new column to existing migrations. Instead you need to create a new migration. 

I think the label `Update preferences table charset` or `Add column team_id in preferences` is what makes the migration unique and are checked by the migration system and there is a `migration_log` table with all the migrations in the DB.

Comment (hugohaggmark) [reply pkg/services/sqlstore/migrations/preferences_mig.go:18]: And this new migration should use the feature toggle of course 😄

File: pkg/services/sqlstore/preferences.go [hunk 2]
```diff
@@ -42,6 +42,7 @@ func (ss *SQLStore) GetPreferencesWithDefaults(ctx context.Context, query *model
 		}
 
 		res := &models.Preferences{
+			NavPosition:     ss.Cfg.NavPosition,
```
Comment (hugohaggmark) [inline pkg/services/sqlstore/preferences.go:45]: All these backend changes need to be behind feature toggle, so we need to figure out how to pass the config.

File: public/app/core/components/sidemenu/BottomSection.tsx [hunk 3]
```diff
@@ -45,7 +46,12 @@ export default function BottomSection() {
   return (
     <div data-testid="bottom-section-items" className={styles.container}>
       {!isSignedIn && (
-        <SideMenuItem label="Sign In" target="_self" url={forcedLoginUrl}>
+        <SideMenuItem
+          label="Sign In"
+          menuPosition={getOppositePosition(navPosition)}
+          target="_self"
+          url={forcedLoginUrl}
+        >
```
Comment (hugohaggmark) [inline public/app/core/components/sidemenu/BottomSection.tsx:54]: Featuretoggle?

**Review by hugohaggmark — COMMENTED** [review]
Substantial progress, I think we need to secure the feature toggle everywhere so we can merge this to main. I can help with the backend code if you want. I haven't tested it yet, we need to fix the migrations first.

One last thing, we need to add tests for all this in the backend.

File: public/app/AppWrapper.tsx [hunk 4]
```diff
@@ -123,3 +146,11 @@ export class AppWrapper extends React.Component<AppWrapperProps, AppWrapperState
     );
   }
 }
+
+const getStyles = (navPosition: typeof contextSrv.user.navPosition) => ({
+  grafanaApp: css`
+    @media only screen and (min-width: 769px) {
```
Comment (ashharrison90) [inline public/app/AppWrapper.tsx:152]: not sure how i can access the theme here, hence the hardcoded value instead of using the theme breakpoint 🤔

Comment (hugohaggmark) [reply public/app/AppWrapper.tsx:152]: You can extend the Props interface with Themable2 and use withTheme2 I think

**Comment by ashharrison90** [comment]
i think we're not sure if we want to introduce this extra complexity, so i'm going to park it and close for now. we'll probably resurrect the branch at a later date to help with mobile support.

File: public/app/AppWrapper.tsx [hunk 5]
```diff
@@ -123,3 +156,13 @@ export class AppWrapper extends React.Component<AppWrapperProps, AppWrapperState
     );
   }
 }
+
+const getStyles = (theme: GrafanaTheme2, navPosition: typeof contextSrv.user.navPosition) => ({
+  grafanaApp: css`
+    @media ${styleMixins.mediaUp(`${theme.breakpoints.values.md}px`)} {
```
Comment (torkelo) [inline public/app/AppWrapper.tsx:162]: the new theme has this built in: https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/themes/GlobalStyles/card.ts#L127

Comment (ashharrison90) [reply public/app/AppWrapper.tsx:162]: ah good to know, there's probably a few more places i can fix then

**Comment by torkelo** [comment]
@ashharrison90 definitely think this is interesting, I am not a big fan of the left icon bar right now so being able to change is interesting :)

**Comment by ashharrison90** [comment]
yeah i like it as well (even if i'd probably stick with the left nav). it's a bit more of demonstrable qol improvement than some of the others we've been making up to now.

think we just need to check what impact it will have on the accessibility improvements we make before deciding to commit to implementing it. it doesn't matter too much as it would always have been hidden behind the feature toggle til 9.0 anyway. 😄

**Pull request closed without merging**